crates/services/parallel-executor/src/validator/dependency_graph.rs

```diff
@@ -0,0 +1,522 @@
+            self.independent_transactions.insert(index);
+            self.newly_independent_queue.push_back(index);
+        }
+    }
+
+    /// Batch add multiple transactions for better performance
+    pub fn add_transactions<I>(&mut self, transactions: I)
+    where
+        I: IntoIterator<Item = (usize, Transaction)>,
+    {
+        for (index, transaction) in transactions {
+            self.add_transaction(index, transaction);
+        }
+    }
+
+    /// Get transactions that are currently independent
+    pub fn get_independent_transactions(&self) -> Vec<usize> {
+        self.independent_transactions.iter().copied().collect()
+    }
+
+    /// Pop next independent transaction from queue
+    pub fn pop_independent_transaction(&mut self) -> Option<usize> {
+        self.newly_independent_queue.pop_front()
+    }
+
+    /// Check if there are any independent transactions ready
+    pub fn has_independent_transactions(&self) -> bool {
+        !self.independent_transactions.is_empty()
+    }
+
+    /// Remove a transaction and update dependencies
+    pub fn remove_transaction(&mut self, tx_index: usize) -> Option<Vec<usize>> {
+        // remove the transaction from our tracking
+        self.transactions.remove(&tx_index)?;
+        let contracts = self.transaction_to_contracts.remove(&tx_index)?;
+        self.remaining_dependencies.remove(&tx_index);
+
+        // remove from independent set if it was there
+        self.independent_transactions.remove(&tx_index);
+
+        let mut newly_independent = Vec::new();
+
+        // pre-collect transactions that might be affected
+        let mut potentially_affected = HashSet::new();
+        for &contract_id in &contracts {
+            if let Some(tx_set) = self.contract_to_transactions.get(&contract_id) {
+                potentially_affected.extend(tx_set.iter().copied());
+            }
+        }
+
+        // remove this transaction from affected contracts
+        for contract_id in contracts {
+            if let Some(tx_set) = self.contract_to_transactions.get_mut(&contract_id) {
+                tx_set.remove(&tx_index);
+
+                if tx_set.is_empty() {
+                    self.contract_to_transactions.remove(&contract_id);
+                }
+            }
+
+            // only check potentially affected transactions
+            for &other_tx_idx in &potentially_affected {
+                if other_tx_idx == tx_index {
+                    continue; // skip the removed transaction
+                }
+
+                if let Some(remaining_deps) =
+                    self.remaining_dependencies.get_mut(&other_tx_idx)
+                {
+                    if let Some(other_contracts) =
+                        self.transaction_to_contracts.get(&other_tx_idx)
+                    {
+                        if other_contracts.contains(&contract_id) {
+                            *remaining_deps -= 1;
+                            if *remaining_deps == 0 {
```

Comment:
delete from `remaning_deps`, `transactions_to_contracts` etc...

Reply:
addressed in 66cdc57f52